### ISSUE TITLE
Default buffering of Event Timing Entries is based on durationThreshold: 104 -- even for Interactions.

### ISSUE THREAD

**Issue by mmocny** [WICG/event-timing#124]
Buffering of EventTiming entries, which buffers from page load until a PerformanceObserver is registered, is always based on a durationThreshold: 104.

When you register a new PerformanceObserver that asks for events with durationThreshold lower than that (i.e. `durationThreshold: 40` to options), there cannot already be any such Entries stored in the buffer.  It may be surprising to receive buffered entries >= 104, but none in the [40, 104) range.

A few options:
* Change spec to buffer certain types of Events with a lower default threshold (i.e. all discrete events, or only those with interactionId)
* Change spec to support "dynamic buffering".  Start by buffering entries with any duration until buffer is full.  Once buffer fills, filter out the fastest entries and bump the threshold.  Keep doing so until threshold reaches 104ms, after which buffer just stays full.

**Comment by yoavweiss** [comment]
Dynamic buffering is interesting! The main concern with buffering "too much" is around memory constraints, and can be a good solution to that. It'd require a bit more work on the UA side of things though, and would require that to happen even on pages where there will never be an EventTiming listener.

I'd be curious to hear from RUM collectors if this is a real issue they are bumping up against when collecting EventTiming entries.

**Comment by mmocny** [comment]
True that buffing, and therefore dynamic buffering, is a cost that is paid on every page even if it ends up not used (by design).  Filtering a fixed sized list on simple criteria, only when its full feels like a small constant amount of work, but its not free.

There were some proposals in the past to support a policy via header/meta tags I think to control overriding this behaviour... Anyway, I don't think dynamic buffering is ideal solution here anyway, because Event Timing reports on so many different event types, and the event counts are not evenly distributed (see below).


> I'd be curious to hear from RUM collectors if this is a real issue they are bumping up against when collecting EventTiming entries.

@philipwalton the author of web-vitals.js library reported this as a real issue.  Users of that library noticed it when trying to use it to measure INP.  (The web-vitals library uses 40ms as the default durationThreshold and this issue means there results of measurement are dependant on the timing of when the library loads-- which PO buffering is meant to prevent.)

But yes, I would love more input here to see if this is a real issue, given that we still do have eventCounts for things under threshold.

---

Try this to show total eventCounts for any page:
```js
console.table([...performance.eventCounts.entries()])
```

Inject early in page load to measure events >=40ms only:
```js
let c = 0;
new PerformanceObserver(list => {
  for (let entry of list.getEntries()) {
    console.log(c++, entry.duration)
  }
}).observe({ type: 'event', durationThreshold: 40 });
```

Anecdote: On my desktop using a mouse pointer, after a mere few seconds on a popular news site I saw reported already 492 events which had duration over 40ms (due to loading main thread contention).  All those events were continuous i.e. pointerenter/leave events or similar.

Dynamic buffering would help, but maybe having different default threshold for events with interactionId is better?

**Comment by philipwalton** [comment]
> @philipwalton the author of web-vitals.js library reported this as a real issue.

My biggest gripe with the current behavior is that it's unexpected.

From an API design perspective, it's not great to have a configuration object where two of the options don't work together and the only way to discover that is to read the spec :)

I understand that the buffering concerns are important, and I'm not suggesting that browsers should buffer all entries as a solution to this. However, I do think there's a real issue here. And what makes matters worse is the issue is not immediately obvious, so a developer could spend a lot of time building and deploying some monitoring code only to realize much latter that they have gaps when looking at the data they collect.

Some potential options to improve this:

- Throw an error if `buffered` and `durationThreshold` are used together. This would mean developers who want both would have to create two POs, but at least they would be aware right away that there's a problem.
- Remove the `durationThreshold` option and make it configurable via a meta tag / header (as Michal suggests).
- Default to buffering all events, but limit the types of events that get buffered by default (or also make that configurable via a meta tag / header).

On that last point, do we know of any sites / RUM providers who are using Event Timing and looking at mouse/pointer events? I wonder if issues related to lots of continuous events would be better handled by a separate API (e.g. something like frame timing).

To comment on some of Michal's suggestions

> A few options:
> 
> * Change spec to buffer certain types of Events with a lower default threshold (i.e. all discrete events, or only those with interactionId)
> * Change spec to support "dynamic buffering".  Start by buffering entries with any duration until buffer is full.  Once buffer fills, filter out the fastest entries and bump the threshold.  Keep doing so until threshold reaches 104ms, after which buffer just stays full.

I think both of these would do a lot to decrease the chances that anyone notices (or is negatively impacted by) this issue, but I don't think they address the root problem.

I wonder if we could use one of these suggestions along with something that makes the API less prone to misuse.

**Comment by nhelfman** [comment]
Just a small input on this. My feeling is that the "dynamic buffering" proposal can be confusing for developers,  even if spec changes have very concrete and detailed definition of the behavior.

**Comment by yoavweiss** [comment]
Alternatively, we currently buffer up to 150 entries before an observer is registered. If we lower the duration for these entries, we're likely to reach that limit sooner, but maybe that's enough for perf observers to be registered?

@mmocny - would you be able to add use counters for cases where perf observers with a buffered flag were registered after the event timing buffer was full? And then maybe experiment with lowering the threshold and seeing if that count increases for those users?

**Comment by mmocny** [comment]
Sure, will file an issue.

---

I will say: my assumption is that this will depend entirely on the hardware used, and the number of continuous events fired.

You can try this on any page you use:

```
function getTotalEventCounts() {
    return Array.from(performance.eventCounts.values()).reduce((a,b) => a+b);
}
setInterval(() => console.log(getTotalEventCounts()), 1000)
```

**Comment by mmocny** [comment]
One more idea:

One of the biggest risks of noisy PerformanceObservers is needles waking of scheduler and calls into v8 (citation needed).  What if we could buffer event timings under 104ms, or under 40ms, but we wouldn't immediately activate the observer for these things?

Right now there is a single shared timer to wake all observers and dispatch all entries, and it has a default delay of 0ms (akin to `setTimeout(0)` after the first entry is ready).  Perhaps we could do something more akin to `requestIdleCallback()` for Event Timings that aren't high priority...

**Comment by andydavies** [comment]
> I'd be curious to hear from RUM collectors if this is a real issue they are bumping up against when collecting EventTiming entries.

@yoavweiss Yes the 104ms default limit is an issue for SpeedCurve and I suspect other RUM providers

Similar to web-vitals.js we use `first-input` entries until our observer for EventTiming entries is registered

Below 104ms our data is dominated by `first-input` entries - `pointerdown`, `mousedown`, `keydown` and so we're probably missing what would be considered the 'real' INP event below 104ms e.g. a click

We may end up having to register an observer and buffer the entries using a inline script as in many cases it's impossible to load a RUM script sooner

RUM needs consent under GDPR and some other privacy regimes so we're often not loaded until that consent is given – in this situation we can end up measuring the very short `pointerdown` on the OK button rather than the actual processing the `click` handler does

**Comment by mmocny** [comment]
Perhaps we could change "first-input" to report more than a single entry, and instead report the "whole first interaction".  Because I agree that the "down" part of the interaction is often less useful, except for the cases of first-input-delay only.

We already want to define that "first-input" is "first event timing with non-0 interactionID" in the spec.

---

Or, as stated early, buffer all "event" timings with interactionID regardless of duration.

Potentially this requires changes to buffering-- using a separate buffer for interactions, or "dynamic" buffering where we filter the buffer of non-interaction events once the buffer is full, or just never buffering non-interaction events.

**Comment by tunetheweb** [comment]
> Perhaps we could change "first-input" to report more than a single entry, and instead report the "whole first interaction". 

I don't think this will really help that much.

Think of the following scenarios:

**Scenario 1 - single interaction before RUM loaded:**

- Click 1:
  - `pointerdown`: 32ms
  - `pointerup`: 132ms
  - `click`: 132ms
- RUM loaded

INP would be reported as 32ms. Your first solution might work for this scenario.

**Scenario 2 - multiple interactions before RUM loaded:**

- Click 1:
  - `pointerdown`: 32ms
  - `pointerup`: 32ms
  - `click`: 32ms
- Click 2:
  - `pointerdown`: 32ms
  - `pointerup`: 132ms
  - `click`: 132ms
- RUM loaded

Your first solution wouldn't work for this.

Given that FID is now gone, and we're really only using `first-input` as a proxy for "missed interactions due to Performance Observer limitations", perhaps we should introduce a `first-input` replacement providing the `longest-interaction`? With a buffer size of 1. Or a buffer size of 10 that allows all events with that `interactionId` to be logged.

That would allow you to get the biggest interaction (which you want for INP), without the overhead of buffering much more.

Alternatively, for an existing solution, you could use LoAF to measure the time between `firstUIEventTimestamp` and the end of the LoAF (or ideally the presentation time if/when that's added to LoAF) to capture any the longest interaction between 50ms and 104ms (though without all the details like node, and INP subparts). And I'd argue that 50ms is likely sufficient for most, especially with a fallback to `first-input` on top of that!

Though to be honest, with the combo of `first-input` (allowing you to include quick INP pages), and 104ms default allows INP to be fairly well calculated, with more detail for interactions above 104ms. But when wanting to chart histogram of interactions to [make a correlation chart for business metrics](https://www.speedcurve.com/blog/site-speed-business-correlation/), then it does expose a clear gap. A gap that either LoAF could fill partially, or `longest-interaction` could fill more completely.
